sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/clienttelemetry/ClientMetricsDiagnosticsHandler.java

```diff
@@ -7,29 +7,102 @@
+    private static final int OPERATION_RECORDING_EXECUTOR_THREAD_KEEP_ALIVE_TIME_IN_MINUTES = 5;
+
     private static final ImplementationBridgeHelpers.CosmosClientTelemetryConfigHelper.CosmosClientTelemetryConfigAccessor clientTelemetryConfigAccessor =
         ImplementationBridgeHelpers.CosmosClientTelemetryConfigHelper.getCosmosClientTelemetryConfigAccessor();
 
     private final CosmosAsyncClient client;
+    private final ExecutorService executorService;
 
-    public ClientMetricsDiagnosticsHandler(CosmosAsyncClient client) {
+    public ClientMetricsDiagnosticsHandler(CosmosAsyncClient client, ConnectionPolicy connectionPolicy) {
 
         checkNotNull(client, "Argument 'client' must not be null.");
+        checkNotNull(connectionPolicy, "Argument 'connectionPolicy' must not be null.");
+
         this.client = client;
+
+        this.executorService = new ThreadPoolExecutor(
+            OPERATION_RECORDING_EXECUTOR_THREAD_MIN_COUNT,
+            OPERATION_RECORDING_EXECUTOR_THREAD_MAX_COUNT,
+            OPERATION_RECORDING_EXECUTOR_THREAD_KEEP_ALIVE_TIME_IN_MINUTES,
+            TimeUnit.MINUTES,
+            new LinkedBlockingQueue<>(this.getOperationRecordingExecutorQueueSize(connectionPolicy)),
+            new RejectedExecutionHandler() {
+                @Override
+                public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+                    if (!executor.isShutdown()) {
+                        Runnable task = executor.getQueue().poll();
+                        executor.execute(r);
+                        // TODO: maybe use debug log level
+                        logger.warn(
+                            "ClientMetricsRecordingQueue is full, dropping metrics reporting for operation {}",
+                            task.toString());
+                    }
+                }
+            }); // if the task queue being full, then discard the oldest task
     }
 
     @Override
     public void handleDiagnostics(CosmosDiagnosticsContext diagnosticsContext, Context traceContext) {
         checkNotNull(traceContext, "Argument 'traceContext' must not be null.");
 
-        ClientTelemetryMetrics.recordOperation(
-            this.client,
-            diagnosticsContext);
+        try {
+            this.executorService.submit(new ClientMetricsOperationRecordingTask(this.client, diagnosticsContext));
+        } catch (Exception e) {
+            // TODO: maybe use debug log level
+            logger.warn("Record metrics failed for operation {} with exception {}", diagnosticsContext.toString(), e);
+        }
+    }
+
+    private int getOperationRecordingExecutorQueueSize(ConnectionPolicy connectionPolicy) {
+        // roughly equal to maxConcurrentRequest for 1000 physical partitions
+        return connectionPolicy.getMaxRequestsPerConnection() * connectionPolicy.getMaxRequestsPerConnection() * 4 * 1000;
```

Comment:
Power of 2 for getMaxRequestsPerConnection, why? - I also think it is definitely safe to have that many entries - you asked whether that might be too high - I think t is a good compromise

Reply:
it calculated by 1000 partitions * 4 replicas * (for each replica addresses, the maximum concurrent requests)

Reply:
oh I see what you mean now, currently changed into use 50K for the queue size